python/dgl/graphbolt/impl/torch_based_feature_store.py

```diff
@@ -12,7 +13,132 @@
+__all__ = ["TorchBasedFeature", "DiskBasedFeature", "TorchBasedFeatureStore"]
+
+
+class DiskBasedFeature(Feature):
+    r"""A wrapper of disk based feature.
+
+    Initialize a disk based feature fetcher by a numpy file.
+
+    Parameters
+    ----------
+    path : string
+        The path to the numpy feature file.
+        Note that the dimension of the numpy should be greater than 1.
+    Examples
+    --------
+    >>> import torch
+    >>> from dgl import graphbolt as gb
+    >>> torch_feat = torch.arange(10).reshape(2, -1)
+    >>> feature = gb.DiskBasedFeature(torch_feat)
+    >>> feature.read(torch.tensor([0]))
+    tensor([[0, 1, 2, 3, 4]])
+    >>> feature.size()
+    torch.Size([5])
+    """
+
+    def __init__(self, path: str, metadata: Dict = None):
+        super().__init__()
+        mmap_mode = "r+"
+        self._tensor = torch.from_numpy(
+            np.load(path, mmap_mode=mmap_mode)
+        ).contiguous()
+
+        self._metadata = metadata
+        self._ondisk_npy_array = torch.ops.graphbolt.ondisk_npy_array(
+            path, self._tensor.dtype, torch.tensor(self._tensor.shape)
+        )
+
+    def read(self, ids: torch.Tensor = None):
+        """Read the feature by index.
+        The returned tensor will be on CPU.
+        Parameters
+        ----------
+        ids : torch.Tensor
+            The index of the feature. Only the specified indices of the
+            feature are read.
+        Returns
+        -------
+        torch.Tensor
+            The read feature.
+        """
+        if ids is None:
+            if self._tensor.is_pinned():
+                return self._tensor.cuda()
+            return self._tensor
+        elif platform.system() == "Linux":
+            try:
+                return self._ondisk_npy_array.index_select(ids.cpu()).to(
+                    ids.device
+                )
+            except RuntimeError:
+                raise IndexError
+        else:
+            return index_select(self._tensor, ids)
+
+    def size(self):
+        """Get the size of the feature.
+        Returns
+        -------
+        torch.Size
+            The size of the feature.
+        """
+        return self._tensor.size()[1:]
+
+    def update(self, value: torch.Tensor, ids: torch.Tensor = None):
+        """Disk based feature does not support update for now."""
+        raise NotImplementedError
+
+    def metadata(self):
+        """Get the metadata of the feature.
+        Returns
+        -------
+        Dict
+            The metadata of the feature.
+        """
+        return (
+            self._metadata if self._metadata is not None else super().metadata()
+        )
+
+    def pin_memory_(self):
+        """Disk based feature does not support pinning memory for now."""
+        raise NotImplementedError
+
+    def is_pinned(self):
+        """Returns True if the stored feature is pinned."""
+        return False
+
+    def to(self, device):  # pylint: disable=invalid-name
```

Comment:
We most likely should not support the `to` function either. Let's clean the code. Would the user want to convert between TorchBasedFeature and DiskBasedFeature?

Reply:
OK, I just want to avoid possible errors of the call, so make it resemble TorchBaseFeature as much as possible. Let's remain them as NotImplement.

Reply:
Where do you get an error? https://docs.dgl.ai/en/latest/generated/dgl.graphbolt.BasicFeatureStore.html#dgl.graphbolt.BasicFeatureStore
BasicFeatureStore does not have pin_memory_ or to methods. If there will be an error because there is no `to` method, then it means we are calling the method. But we will get an error anyway because it is not implemented. So, I don't see the benefit of keeping these methods. Could you elaborate?

Reply:
At the very beginning, I find it will occur errors if we direct apply DiskFeature to run this example https://github.com/dmlc/dgl/blob/master/examples/sampling/graphbolt/rgcn/hetero_rgcn.py#L573, for small dataset, it should be pinned on memory, but for large dataset it's impossible. I want Disk Feature to keep the ability to deal with small datasets (e.i. the same as TorchBased). However, if we want to split it as extra method to only deal with large datasets, we can throw all useless function.

Reply:
I think DiskBasedFeature's APIs should be kept consistent with TorchBasedFeature, as we consider the two classes to be equivalent.

Reply:
Why do we consider them equivalent? One is a torch tensor, the other is tensor data stored on the disk.

Reply:
We can instead write a method to get a TorchBasedFeature out of a DiskBasedFeature.

Reply:
read_into_ram() -> TorchBasedFeature

Something like above.

Reply:
Because we want to replace the mmap-based method in the TorchBasedFeatureStore at the beginning: https://github.com/dmlc/dgl/blob/master/python/dgl/graphbolt/impl/torch_based_feature_store.py#L287

Reply:
Then we can modify TorchBasedFeatureStore's to and pin_memory_ methods so that it skips over DiskBasedFeature s in it.

Reply:
OK, it is a possible solution. let's do as you say.

Reply:
If we have read_into_memory, then we could simply implement `.to` as:

```
def to(self, device): -> TorchBasedFeature
    return self.read_into_memory().to(device)
```

Reply:
We can implement `.to` as above but `pin_memory_` is a bit complicated because it is an in-place operation. We can't convert a DiskBasedFeature to a TorchBasedFeature in-place.